pkg/scheduler/core/equivalence/eqivalence.go

```diff
@@ -150,6 +140,25 @@ func (c *Cache) InvalidateCachedPredicateItemForPodAdd(pod *v1.Pod, nodeName str
 		}
 	}
 	c.InvalidatePredicatesOnNode(nodeName, invalidPredicates)
+
+	// MatchInterPodAffinity: we assume scheduler can make sure newly bound pod
+	// will not break the existing inter pod affinity. So normally we does not need to
+	// invalidate MatchInterPodAffinity when pod added.
+	//
+	// NOTE: assumptions above will not stand when we implemented features like
+	// RequiredDuringSchedulingRequiredDuringExecutioc.
+	//
+	// But the exception is: pod has anti-affinity to itself. This is common in a ReplicaSet,
+	// so we have to invalidate MatachInterPodAffinity in this case.
+	//
+	// TODO(harry): This is aggressive. We should consider a more fine-grained way later,
+	// especially after eCache is merged to predicates pkg (#67241). Then we can do invalidation:
+	// 1. only for nodes in the same failure domain;
+	// 2. only for pods with self anti-affinity.
+	invalidAntiAffinityPred := sets.NewString(predicates.MatchInterPodAffinityPred)
+	if pod.Spec.Affinity != nil && pod.Spec.Affinity.PodAntiAffinity != nil {
```

Comment:
Any other pod could have an affinity or anti-affinity rule that matches this pod, even if there are no affinity rules on the new pod.

Reply:
Emm, for a newly added pod without affinity rule defined, it either went thru real predicates, or reused cached predicates of an existing pod. It seems in both cases, the affinity/anti-affinity assumption of this node (and this failure domain) will not be broke? Any exception in your mind?

The main concern here is we have to invalidate predicates for all nodes. But fine-grained control is only possible when we can access `v1.Node` in ecache.

Thus, I recalled you've also mentioned in #67241, I would suggest let's combine the cheap predicates logic in #67241 to this PR, and add `MatchInterPodAffinity` into the skip list temporarily. So we can ship these two fixes in this cycle and continue merging eCache into predicates work at the same time. cc @k82cn.  

Sounds reasonable?

Reply:
An example that demonstrates Jonathan's point could be something like:

1. Pod A has anti-affinity to Pods with label "foo"
2. Pod A ran through predicates once, but hit some error somewhere
3. Pod B comes along with label "foo" (but no pod.Spec.Affinity specified) and gets scheduled on node X
4. The anti-affinity predicate for Pod A should be invalidated on node X

If this was affinity instead, the the affinity predicate should be invalidated on all other nodes except X (in the optimized case)

Reply:
Thanks Michelle. This is actually a independent case I'm trying to figure out recently. i.e. When a Pod went thru predicates but did not occur on the node. We should invalidate all predicates (or specific predicates at least) for that Pod equiv class.

Right now the place I can think of doing this is after assume() returns error. Not sure if there's other "hit some error somewhere" case in your mind?

Reply:
Prefer to skip `MatchInterPodAffinity` for eCache instead of covering it case by case :(

> Any exception in your mind?

Affinity itself :), If no pod on any host, then all nodes are fits; after that, pod affinity only
schedule to the node with that pod.

Reply:
> Affinity itself :), If no pod on any host, then all nodes are fits; after that, pod affinity only
schedule to the node with that pod.

eCache is keyed by node, I don't think "all nodes are fits" is possible ... mind explain a bit?

My proposal is temporarily skip this predicate and other cheap ones as well. Fine-grained control can be shipped after we achieved predicates pkg refactoring.

Reply:
> Affinity itself

If no pod on host_1 nor host_2, the predicates of both hosts are fit; after one pod dispatches to host_1, only host_1 should be fits. For eCache, if pod fit host_1 and host_2, there'll be two record (keyed by node); but it'll not refresh it.

> Not sure if there's other "hit some error somewhere" case in your mind?

Can we stop list cases here? IMO, Skip `MatchInterPodAffinity` is safer for this release.

Reply:
> If no pod on host_1 nor host_2, the predicates of both hosts are fit; after one pod dispatches to host_1, only host_1 should be fits. For eCache, if pod fit host_1 and host_2, there'll be two record (keyed by node); but it'll not refresh it.

eCache is updated when pod is scheduled  to corresponding node. i.e. if `one pod dispatches to host_1`, then only `eCache[host_1]` is set, not sure why `there'll be two record (keyed by node)`?

> Can we stop list cases here?

Sorry, but the context of that part of discussion is about an exception affects ALL predicates, i.e. when sth wrong happened before bind(), we will have stale eCache. Also see this re-scheduling comment: https://github.com/kubernetes/kubernetes/issues/67680#issuecomment-415635647 , I have a patch for this fix.

> Skip MatchInterPodAffinity is safer for this release.

I can confirm this conclusion has been reached.

Reply:
Is PodAdd also invoked when a pod is created? Not just when a pod is assumed?

Reply:
> eCache is updated when pod is scheduled to corresponding node. i.e. if one pod dispatches to host_1, then only eCache[host_1] is set

Sorry, did not check the detail of code. So only one node is ok for the eq pods? will scheduler dispatch all eq pods to the same node? or only the predicates on this node will not re-run?

Reply:
Just re-use the predicates results of this node.

Reply:
@msau42 Yes, but we have filter and separate handlers for assigned/unassigned Pods. The scheduler cache set & eCache invalidation are all handled by assigned Pod handler.